examples/ufo_toy.rs

```diff
@@ -0,0 +1,164 @@
+use druid_shell::win_main;
+
+use druid::widget::Widget;
+use druid::{
+    BoxConstraints, Geometry, Id, LayoutCtx, LayoutResult, PaintCtx,
+    UiMain, UiState,
+};
+
+struct GlyphEditor {
+    glyph: Glyph,
+    path: BezPath,
+    height: f32,
+}
+
+impl GlyphEditor {
+    fn new(glyph: Glyph) -> Self {
+        // assume glyph height is 1000 or 2000 units
+        let height = if glyph.outline.as_ref()
+            .map(|o| o.contours.iter()
+                 .flat_map(|c| c.points.iter().map(|p| p.y))
+                 .any(|h| h > 1000.))
+            .unwrap_or(false) { 4000. } else { 1000. };
+
+        let path = glyph.outline.as_ref().map(|o| make_path(&o.contours)).unwrap_or_default();
+        GlyphEditor { glyph, height, path }
+    }
+}
+
+impl Widget for GlyphEditor {
+    fn paint(&mut self, ctx: &mut PaintCtx, geom: &Geometry) {
+
+        let baseline = (geom.size.1  * 0.66) as f64;
+        let l_pad = 100.;
+
+        let baseline_fg = ctx.render_ctx.solid_brush(0x00_80_f0_ff).unwrap();
+        let fg = ctx.render_ctx.solid_brush(0xf0_f0_ea_ff).unwrap();
+        let control_point_fg = ctx.render_ctx.solid_brush(0x70_80_7a_ff).unwrap();
+
+        let scale = ( geom.size.1 / self.height * 0.65).min(1.0).max(0.2);
+        println!("scale {}", scale);
+        let affine = Affine::new([scale as f64, 0.0, 0.0, -scale as f64, l_pad, baseline]);
+
+        let line = Line::new((0., baseline), (geom.size.0 as f64, baseline));
+        ctx.render_ctx.stroke(line, &baseline_fg, 0.5, None);
+
+
+        ctx.render_ctx.stroke(affine * &self.path, &fg, 1.0, None);
+
+        for shape in self.glyph.outline.as_ref().iter().map(|o| o.contours.iter()).flatten() {
+            for point in shape.points.iter() {
+                println!("{:?}", point);
+                let color = if point.typ == PointType::OffCurve { &control_point_fg } else { &fg };
+                let circ = Circle::new((l_pad + (point.x * scale) as f64, (baseline - (point.y * scale) as f64)), 8.0 * scale as f64);
+                ctx.render_ctx.fill(circ, color, FillRule::NonZero);
+            }
+        }
+    }
+
+    fn layout(
+        &mut self,
+        bc: &BoxConstraints,
+        _children: &[Id],
+        _size: Option<(f32, f32)>,
+        _ctx: &mut LayoutCtx,
+    ) -> LayoutResult {
+        LayoutResult::Size(bc.constrain((100.0, 100.0)))
+    }
+}
+
+fn main() {
+    let glyph_path = match std::env::args().skip(1).next() {
+        Some(arg) => arg,
+        None => {
+            eprintln!("Please pass a path to a .glif file");
+            std::process::exit(1);
+        }
+    };
+
+    println!("loading {}", glyph_path);
+    let glyph = norad::Glyph::load(&glyph_path).expect("failed to load glyph");
+
+    druid_shell::init();
+
+    let mut run_loop = win_main::RunLoop::new();
+    let mut builder = WindowBuilder::new();
+    let mut state = UiState::new();
+
+    let root = GlyphEditor::new(glyph);
+    let root_id = state.add(root, &[]);
```

Comment:
This works but I try to have a `.ui` method on widgets that does this. Among other things, that gives you slightly more flexibility to use a builder pattern where it actually adds multiple lower-level widgets to the tree. But this is something that maybe should be rethought, especially because it treats initial construction differently than dynamically adding stuff later.

Reply:
I saw that pattern, but it wasn't totally clear what the rationale was, so I wanted to just do the simplest, most comprehensible thing (to me). This is my first dabble with druid, so definitely not up to speed on patterns etc.